core/events/events_abstract.js

```diff
@@ -24,98 +24,107 @@ const {Workspace} = goog.requireType('Blockly.Workspace');
 
 /**
  * Abstract class for an event.
- * @constructor
- * @alias Blockly.Events.Abstract
  */
-const Abstract = function() {
+class Abstract {
   /**
-   * Whether or not the event is blank (to be populated by fromJson).
-   * @type {?boolean}
+   * @alias Blockly.Events.Abstract
    */
-  this.isBlank = null;
+  constructor() {
+    /**
+     * Whether or not the event is blank (to be populated by fromJson).
+     * @type {?boolean}
+     */
+    this.isBlank = null;
+
+    /**
+     * The workspace identifier for this event.
+     * @type {string|undefined}
+     */
+    this.workspaceId = undefined;
+
+    /**
+     * The event group id for the group this event belongs to. Groups define
+     * events that should be treated as an single action from the user's
+     * perspective, and should be undone together.
+     * @type {string}
+     */
+    this.group = eventUtils.getGroup();
+
+    /**
+     * Sets whether the event should be added to the undo stack.
+     * @type {boolean}
+     */
+    this.recordUndo = eventUtils.getRecordUndo();
+
+    /**
+     * Whether or not the event is a UI event.
+     * @type {boolean}
+     */
+    this.isUiEvent = false;
+
+    /**
+     * Type of this event.
+     * @type {string|undefined}
+     */
+    this.type = undefined;
```

Comment:
Just want to note, I'm pretty sure we let external peeps define events (I think AppInventor does this?)

Reply:
I think we could reasonably go to a string and have the empty string if not set--but when I tried that assorted tests broke, mostly because of assertions about what the type property was, so I left it alone in the interest of making a minimal change. I expect to return to this to address typescript compiler warnings about this value possibly being undefined.